master/internal/api_experiment.go

```diff
@@ -28,46 +28,49 @@ import (
 	"github.com/pkg/errors"
 
 	"github.com/determined-ai/determined/master/internal/db"
+	expauth "github.com/determined-ai/determined/master/internal/experiment"
 	"github.com/determined-ai/determined/master/internal/grpcutil"
 	"github.com/determined-ai/determined/master/internal/hpimportance"
 	"github.com/determined-ai/determined/master/internal/lttb"
 	"github.com/determined-ai/determined/master/pkg/actor"
 	"github.com/determined-ai/determined/master/pkg/model"
 	"github.com/determined-ai/determined/master/pkg/protoutils"
 	"github.com/determined-ai/determined/master/pkg/protoutils/protoless"
+	"github.com/determined-ai/determined/master/pkg/ptrs"
 	"github.com/determined-ai/determined/master/pkg/schemas"
 	"github.com/determined-ai/determined/master/pkg/schemas/expconf"
 	"github.com/determined-ai/determined/master/pkg/searcher"
 	"github.com/determined-ai/determined/proto/pkg/apiv1"
 	"github.com/determined-ai/determined/proto/pkg/checkpointv1"
 	"github.com/determined-ai/determined/proto/pkg/experimentv1"
 	"github.com/determined-ai/determined/proto/pkg/jobv1"
+	"github.com/determined-ai/determined/proto/pkg/projectv1"
 
 	structpb "github.com/golang/protobuf/ptypes/struct"
 )
 
 var experimentsAddr = actor.Addr("experiments")
 
-func (a *apiServer) checkExperimentExists(id int) error {
-	ok, err := a.m.db.CheckExperimentExists(id)
-	switch {
-	case err != nil:
-		return status.Errorf(codes.Internal, "failed to check if experiment exists: %s", err)
-	case !ok:
-		return status.Errorf(codes.NotFound, "experiment %d not found", id)
-	default:
-		return nil
+func (a *apiServer) getExperiment(
+	curUser model.User, experimentID int,
+) (*experimentv1.Experiment, error) {
+	expNotFound := status.Errorf(codes.NotFound, "experiment not found: %d", experimentID)
+	exp := &experimentv1.Experiment{}
+	if err := a.m.db.QueryProto("get_experiment", exp, experimentID); errors.Is(err, db.ErrNotFound) {
+		return nil, expNotFound
+	} else if err != nil {
+		return nil, errors.Wrapf(err, "error fetching experiment from database: %d", experimentID)
 	}
-}
 
-func (a *apiServer) getExperiment(experimentID int) (*experimentv1.Experiment, error) {
-	exp := &experimentv1.Experiment{}
-	switch err := a.m.db.QueryProto("get_experiment", exp, experimentID); {
-	case err == db.ErrNotFound:
-		return nil, status.Errorf(codes.NotFound, "experiment not found: %d", experimentID)
-	case err != nil:
-		return nil, errors.Wrapf(err,
-			"error fetching experiment from database: %d", experimentID)
+	modelExp, err := model.ExperimentFromProto(exp)
+	if err != nil {
+		return nil, err
+	}
+	if ok, err := expauth.AuthZProvider.Get().
+		CanGetExperiment(curUser, modelExp); err != nil {
```

Comment:
idea: maybe `CanGetExperiment` can take `model.ExperimentId` as the 2nd parameter. this allows to check the permission first, and if that succeeds fetch the experiment.

Reply:
My viewpoint is that I want to avoid as much queries / stuff in the ```AuthZ``` implementations so that we don't need to implement the same getting experiment logic across both implementations.